### PR TITLE
Fix tls.py assertion issues.

### PR DESCRIPTION
**Pull request by rthalley**
1) Some assertions had side effects and would cause a loss of
   framing if python was run with -O or -OO

2) Some assertions should have been error checks.

### CODE REVIEW

**Comment by rthalley** [comment]
I haven't figured out how to run the tests with -O yet, so consider this a draft.  Everything passes for me without -O, and the problem my test server was having when run with -O went away.  Arguably it would be good to add test coverage for the various kinds of input problems that can now cause TLS alerts, but I haven't figured out how to do those yet.

File: src/aioquic/tls.py [hunk 1]
```diff
@@ -365,7 +369,9 @@ def pull_block(buf: Buffer, capacity: int) -> Generator:
     length = int.from_bytes(buf.pull_bytes(capacity), byteorder="big")
     end = buf.tell() + length
     yield length
-    assert buf.tell() == end
+    if buf.tell() != end:
+        # There was trailing garbage or our parsing was bad.
+        raise AlertDecodeError("extra bytes at the end of a block")
```
Comment (jlaine) [inline src/aioquic/tls.py:374]: I think this should remain an assertion. It means the caller of `pull_block` did not read the right number of bytes within the context manager.

File: src/aioquic/tls.py [hunk 2]
```diff
@@ -551,7 +570,12 @@ def pull_extension() -> None:
                 hello.psk_key_exchange_modes = pull_list(buf, 1, buf.pull_uint8)
             elif extension_type == ExtensionType.SERVER_NAME:
                 with pull_block(buf, 2):
-                    assert buf.pull_uint8() == 0
+                    name_type = buf.pull_uint8()
+                    if name_type != 0:
+                        # We don't know this name_type.
+                        raise AlertIllegalParameter(
+                            f"unknown server name type {name_type}"
+                        )
```
Comment (jlaine) [inline src/aioquic/tls.py:578]: I would suggest adding a `pull_server_name` method to facilitate testing. For symmetry you might also consider adding `push_server_name` too.

Comment (jlaine) [reply src/aioquic/tls.py]: .. and done

File: src/aioquic/tls.py (hunk 1 above)
Comment (jlaine) [reply src/aioquic/tls.py:378]: No scratch that comment.

**Pull request merged**